### ISSUE TITLE
Using minetest.generate_decorations() and biome-filtered decos when no biomemap exists, possible bug and a feature request

### ISSUE THREAD

**Issue by LandonAConway** [minetest/minetest#10250]
##### Minetest version

```

```

##### OS / Hardware
Operating system: Win10
CPU: Unknown

##### Summary
When using minetest.generate_decorations(), the method does not generate all decorations known to a biome. For some biomes, all decorations are generated, but, for other biomes such as 'jungle' in 'minetest_game', jungle trees are not generated. Only jungle grass.

##### Steps to reproduce
Create a mod using minetest.generate_decorations() and generate the decorations inside a ranforest area with no trees or jungle grass. Notice how only jungle grass is generated. Then, do the same thing woth the node default:dirt_with_coniferous_litter and notice how it generates pine trees there and does the same for grassland.

**Comment by paramat** [comment]
This depends very much on the code you are using, so please could you link to the exact code that reproduces the bug?
Your 'steps to reproduce' are not specific enough and are open to misinterpretation.
For example, is this code being run during mapgen in `register_on_generated` or not?

My initial guess is that `minetest.generate_decorations()` is not being used during mapgen in `register_on_generated`, so the biomemap does not exist, which might cause biome filtering to fail.

Generally, if you use `minetest.generate_decorations()` you cannot necessarily use a games' standard decoration registrations, they might need changing to work because the decoration-placing situation is so different.

I have used `minetest.generate_decorations()` 4 times in my own mods or games, in all cases it is used to place decorations in a Lua mapgen (this is the intended usage of the feature), obviously there are no engine biomes or a biomemap, so registered decorations do not use the `biome` field.

**Comment by paramat** [comment]
Actually, one of my uses of `minetest.generate_decorations()` is with normal mapgen:

Engine decoration placement is disabled. In `register_on_generated()` Lua code adds structures to the terrain, then  `minetest.generate_decorations()` is used to place the decorations afterwards. I use the `biomes` field in the decoration registration and it works correctly, because the biomemap is present during the execution of `register_on_generated()`.

Please could you test by using `minetest.generate_decorations()` in `register_on_generated()`? And report back if it works or not.

Perhaps we could add documentation to warn of the missing biomemap if using this outside `register_on_generated()`.

**Comment by LandonAConway** [comment]
Yes, I'm using it in an abm to re-generate forests. I wish there was a proper way to regenerate decorations since the method isn't obviosly meant for that. Maybe we should open a new issue for that request or change the title of this issue.

**Comment by paramat** [comment]
Ok, thanks. I changed the title and labels, now a feature request.
As your usage is a minor and unintended usage i am unsure it is justified to calculate biome noises and create a biomemap when `minetest.generate_decorations()` is used outside `register_on_generated()`, but i will consider it.
Maybe we can do something simple to make the 'biome' field of decorations fail in a less disruptive way.

**Comment by sfan5** [comment]
Is this limitation documented? If not, it should be.

**Comment by paramat** [comment]
> the code already skips biome checks when there is no biomemap:

Good.

There is not necessarily a bug here. It is expected that the decoration registrations that rely on biome filtering will fail when there is no biomemap.
The issue author needs to keep this in mind and test again to check that the behaviour is as expected in this situation.

Besides this, calculating the biomemap when it does not exist has been requested above, so this is also a feature request.

**Comment by sfan5** [comment]
The initial report was a bug report saying that biome filtering is broken (≠ will not filter) when there is no biomemap.
Based on your guess that it is indeed broken this issue was changed into a feature request.

It needs to be clarified whether it is indeed broken and only then does it make sense to turn this issue into a feature request.

**Comment by paramat** [comment]
Seems to be some confusion here.
> The initial report was a bug report saying that biome filtering is broken (≠ will not filter) when there is no biomemap.

It is not certain there is a bug here, considering the complexity of MTG decorations the behaviour the author experiences may be expected and explained by the fact that biome filtering is disabled when there is no biomemap.

> Based on your guess that it is indeed broken ...

I did not guess there is a bug or anything is broken.
> ... this issue was changed into a feature request.

That is not why it was changed to a feature request.
> It needs to be clarified whether it is indeed broken and only then does it make sense to turn this issue into a feature request.

Not so, changing the abilities of the feature is the feature request already made in https://github.com/minetest/minetest/issues/10250#issuecomment-668179575
> I wish there was a proper way to regenerate decorations

Which requires calculating the biomemap if `minetest.generate_decorations()` is used in a core mapgen but when the biomemap does not exist (Lua mapgens cannot have an engine biomemap).
Which i think is low priority and probably not worth doing.
This would be necessary whether there is a bug or not.

My last change of the title was done to make it general to include the potential bug and the feature request.
Your new title excludes any possible bug and makes it a feature request only.
